### PR TITLE
Feature/simplify

### PR DESCRIPTION
**Pull request by luiz290788**
# Description
Adds simplify method that simplifies the condition with already existing context.

### Tasks
- [x] add more tests
- [x] add docs
- [x] add examples
- [x] update readme

### Test Coverage
- [x] unit

### CODE REVIEW

File: src/__test__/helpers.ts [hunk 1]
```diff
@@ -30,10 +31,18 @@ export function permutation(inputs: any[]): [any, any][] {
...
-  }
+export const operand = (value: any): Operand => new class extends Operand {
+  type: EvaluableType = EvaluableType.Operand
+  constructor(private readonly value: any) { super() }
   evaluate() { return this.value }
+  simplify() { return this.value }
+  serialize(): Input { throw new Error('not implemented') }
 }(value)
+
+export const notSimplified = (): Operand => new class extends Operand {
```
Comment (rthreei) [inline src/__test__/helpers.ts:42]: What's meant by `notSimplified`?

File: src/common/evaluable.ts [hunk 2]
```diff
@@ -46,6 +49,10 @@ export interface Evaluable {
    */
   evaluate(ctx: Context): Result;
 
+  simplify(ctx: Context): Result | Evaluable
```
Comment (rthreei) [inline src/common/evaluable.ts:52]: Missing trailing `;` -- not sure if that matters, but for consistency.

File: src/expression/comparison/__test__/unit/eq.test.ts [hunk 3]
```diff
@@ -15,24 +15,44 @@ describe('Expression - Comparison - Equal', () => {
     })
   })
 
+  const primitives = [1, '1', true, false, undefined, null]
+  const testCases: [Operand, Operand, boolean][] = [
+    // Truthy cases - type A === type A
+    ...primitives.map<[Operand, Operand, boolean]>((value) => [operand(value), operand(value), true]),
+    // Falsy - different types - across all permutations
+    ...permutation(primitives)
+      .map<[Operand, Operand, boolean]>(([left, right]) => [operand(left), operand(right), false]),
```
Comment (rthreei) [inline src/expression/comparison/__test__/unit/eq.test.ts:24]: Neat. I'm surprised you need to `<[Operand, Operand, boolean]>`?

File: src/expression/logical/and.ts [hunk 4]
```diff
@@ -37,4 +38,29 @@ export class And extends Logical {
     }
     return true
   }
+
+  simplify (ctx: Context): boolean | Evaluable {
```
Comment (rthreei) [inline src/expression/logical/and.ts:42]: Why would `simplify` return `boolean`?

File: src/expression/logical/xor.ts [hunk 5]
```diff
@@ -50,4 +53,37 @@ export class Xor extends Logical {
...
+    if (evaluablesLeft.length === 1) {
+      if (trueCount === 1) {
+        return new Not(...evaluablesLeft)
+      }
+      return evaluablesLeft[0]
+    }
+    if (trueCount === 1) {
+      return new Nor(evaluablesLeft)
+    }
+    return new Xor(evaluablesLeft)
```
Comment (rthreei) [inline src/expression/logical/xor.ts:87]: Do you not prefer whitespace for visual separation? Asking out of curiosity.  Not asking as a prompt to change this.

File: types/common/evaluable.d.ts [hunk 6]
```diff
@@ -17,7 +19,7 @@ export interface Context {
 /**
  * Evaluation result
  */
-export declare type Result = undefined | null | string | number | boolean | Array<string | number | boolean | null>;
+export declare type Result = undefined | null | string | number | boolean | Array<Result>;
```
Comment (rthreei) [inline types/common/evaluable.d.ts:22]: 👍

**Review by rthreei — COMMENTED** [review]
Nice cleanup along the way 👍

File: src/expression/logical/and.ts (hunk 4 above)
Comment (rthreei) [reply src/expression/logical/and.ts:42]: I think I get it now.  I find it strange when a method can return more than one type. I've seen this done a tuple:
`[boolean, Evaluable]`

File: src/__test__/helpers.ts (hunk 1 above)
Comment (davidhorak) [inline src/__test__/helpers.ts:42]: any concerns about just passing: simply: boolean into the operand, to avoid that secondary helper?
```
export const operand = (value: any, simply = false): Operand => new class extends Operand {}
```

File: src/common/evaluable.ts (hunk 2 above)
Comment (davidhorak) [reply src/common/evaluable.ts:52]: I would say let's remove the ```;``` it may a legacy stuff

Comment (davidhorak) [inline src/common/evaluable.ts:52]: Could we drop here a comment, what exactly is the "simplify" method desired outcome? To help others to quickly get the idea of the intention

File: src/__test__/helpers.ts (hunk 1 above)
Comment (luiz290788) [reply src/__test__/helpers.ts:42]: It is just to simulate a `Operand` that was not simplified. This is useful to test cases where you have nested `Evaluable` that could not be simplified forcing the parent not to be simplified.

File: src/expression/comparison/__test__/unit/eq.test.ts (hunk 3 above)
Comment (davidhorak) [reply src/expression/comparison/__test__/unit/eq.test.ts:24]: because this return from map ```=> [operand(value), operand(value), true]``` = ```[Operand|boolean][] type for TS``` where we are more concrete here, we want to the exact tuple array for test cases

Comment (luiz290788) [reply src/expression/comparison/__test__/unit/eq.test.ts:24]: You need. Typescript assumes that the return type is `(Operand | boolean)[]`. It doesn't infer that this is a tuple.

File: src/expression/comparison/eq.ts [hunk 7]
```diff
@@ -23,15 +23,10 @@ export class Equal extends Comparison {
...
-  /**
-   * Evaluate in the given context.
-   * @param {Context} ctx
-   * @return {Result}
-   */
-  evaluate (ctx: Context): Result {
-    return this.left.evaluate(ctx) === this.right.evaluate(ctx)
+  comparison (left: Result, right: Result): boolean {
+    return left === right
   }
```
Comment (davidhorak) [inline src/expression/comparison/eq.ts:31]: why this change?
How this would work: ```[=, $Firstname, "Peter"]```?

Comment (davidhorak) [reply src/expression/comparison/eq.ts:31]: Oh I see, you've generalized the evaluation here: https://github.com/briza-insurance/illogical/blob/3b112351c786352cd5d4c4145af8bd51e294f8d9/src/expression/comparison/index.ts#L35-L37 ... make sense 👍

File: src/expression/logical/and.ts (hunk 4 above)
Comment (luiz290788) [reply src/expression/logical/and.ts:42]: In this case it is not a tuple. In functional programming this would be an either. And if were a tuple it would need to be `[boolean | undefined, Evaluable | undefined]`

File: src/expression/logical/xor.ts (hunk 5 above)
Comment (luiz290788) [reply src/expression/logical/xor.ts:87]: I don't mind the whitespace. I prefer the code tight and small but up to you.

File: src/__test__/helpers.ts (hunk 1 above)
Comment (luiz290788) [reply src/__test__/helpers.ts:42]: Yeah, we can do that.

File: src/common/evaluable.ts (hunk 2 above)
Comment (luiz290788) [reply src/common/evaluable.ts:52]: I will. It is part of my docs task.

File: src/expression/comparison/eq.ts (hunk 7 above)
Comment (luiz290788) [reply src/expression/comparison/eq.ts:31]: Because we can reuse the comparison in the evaluate and simplify.

The parent implements evaluate that evaluates left and right passing them to comparison and returning its value. simplify will try to simplify left and right and if possible call comparison.

File: src/expression/logical/and.ts (hunk 4 above)
Comment (davidhorak) [reply src/expression/logical/and.ts:42]: I do not believe so that the intention here is to return a tuple. When the logical expression is simplified, it may product the final evaluated state, i.e. true|false. In other cases, it returns the expression.

Saying that, I will drop a note about comments for all new functions, as this is a public repo, we should document thoroughly

File: src/common/evaluable.ts (hunk 2 above)
Comment (davidhorak) [reply src/common/evaluable.ts:52]: I missed the TODO :)

File: src/operand/reference.ts [hunk 8]
```diff
@@ -100,6 +101,18 @@ export class Reference extends Operand {
...
+  simplify (ctx: Context): Result | Evaluable {
+    const key = this.key.split(/\./)[0]
+    if (ctx[key] !== undefined) { // TODO excluded ids
+      return this.evaluate(ctx)
+    }
+    return this
+  }
+
+  serialize ({ referenceSerialization }: Options): string {
+    return referenceSerialization(this.key)
```
Comment (luiz290788) [inline src/operand/reference.ts:113]: OMG this is much much much more complex than just this.

**Comment by davidhorak** [comment]
Looking good, nice clean up!

For the documentation, we can drop the ```@return``` statements, and the types e.g.: ```{Context}``` ... these were for JSDoc, but this repo uses now the TSDoc
```
  /**
   * Evaluate in the given context.
   * @param {Context} ctx
   * @return {Result}
   *
```

>>

```
  /**
   * Evaluate in the given context.
   * @param ctx
   *
```

or, where there is no additional intel for a param >>
```
```
  /**
   * Evaluate in the given context.
   *
```

File: src/operand/reference.ts (hunk 8 above)
Comment (davidhorak) [reply src/operand/reference.ts:113]: if needed some insight, @lilyvanoekel would be the best resources, as she's implemented the nested references, composite references.

File: src/expression/logical/xor.ts (hunk 5 above)
Comment (rthreei) [reply src/expression/logical/xor.ts:87]: Terse code.

No no.  I'm not going to request these sort of changes.  Everyone just needs to understand that code should be readable for both peers and yourself.

**Comment by luiz290788** [comment]
This PR is getting bigger and bigger =(

File: src/expression/logical/xor.ts (hunk 5 above)
Comment (luiz290788) [reply src/expression/logical/xor.ts:90]: Do you think the code is unreadable because there is not empty line after the if block?

Comment (rthreei) [reply src/expression/logical/xor.ts:90]: It is readable.

File: src/expression/logical/and.ts (hunk 4 above)
Comment (rthreei) [reply src/expression/logical/and.ts]: Oh I see now.  Right.  Thank you for the explanation.

**Comment by rthreei** [comment]
Looks good to me.  This PR introduces a public interface method `simplify`, which seems reasonable.

David, good to merge and deploy?

**Review by davidhorak — APPROVED** [review]
Looks good not,but let's not forgot about the "changelog.md" and packages version bump:
https://docs.google.com/document/d/19LtdshUTZPLfuYkDv-qYnlObs5J2FaRGEYbLZTGt7UU

**Comment by luiz290788** [comment]
> Looks good not,but let's not forgot about the "changelog.md" and packages version bump:
> https://docs.google.com/document/d/19LtdshUTZPLfuYkDv-qYnlObs5J2FaRGEYbLZTGt7UU

Sure, I was going to ask you about this now. The process is that I update everything (changelog and version) in this branch and then merge and deploy from master?

**Comment by davidhorak** [comment]
> > Looks good not,but let's not forgot about the "changelog.md" and packages version bump:
> > https://docs.google.com/document/d/19LtdshUTZPLfuYkDv-qYnlObs5J2FaRGEYbLZTGt7UU
> 
> Sure, I was going to ask you about this now. The process is that I update everything (changelog and version) in this branch and then merge and deploy from master?

Yeah, as this really represents a feature which is immediately released, it make sense to bump the package here, and update the changelog accordingly.

**Comment by luiz290788** [comment]
> > > Looks good not,but let's not forgot about the "changelog.md" and packages version bump:
> > > https://docs.google.com/document/d/19LtdshUTZPLfuYkDv-qYnlObs5J2FaRGEYbLZTGt7UU
> > 
> > 
> > Sure, I was going to ask you about this now. The process is that I update everything (changelog and version) in this branch and then merge and deploy from master?
> 
> Yeah, as this really represents a feature which is immediately released, it make sense to bump the package here, and update the changelog accordingly.

new version: 1.3.0

It this a minor version bump for you too?

**Comment by davidhorak** [comment]
> > > > Looks good not,but let's not forgot about the "changelog.md" and packages version bump:
> > > > https://docs.google.com/document/d/19LtdshUTZPLfuYkDv-qYnlObs5J2FaRGEYbLZTGt7UU
> > > 
> > > 
> > > Sure, I was going to ask you about this now. The process is that I update everything (changelog and version) in this branch and then merge and deploy from master?
> > 
> > 
> > Yeah, as this really represents a feature which is immediately released, it make sense to bump the package here, and update the changelog accordingly.
> 
> new version: 1.3.0
> 
> It this a minor version bump for you too?

1.3.0 sounds appropriate, as this is really a new "bigger" feature.
